File: 21Jan2024.py
# Problem Link :- https://leetcode.com/problems/house-robber/

from ast import List


class Solution:
    def f(self,a : List[int],i :int, j: int,dp : List[List[int]]) -> int :
        if(i>=len(a)) :
            return 0
        
        if(dp[i][j]!=-1) :
            return dp[i][j]
        f1 = self.f(a,i+1,0,dp)
        ans=f1
        if(j==0) :
            f2 = self.f(a,i+1,1,dp)+a[i]
            ans=max(ans,f2)
        dp[i][j]=ans
        return ans
        
    def rob(self, nums: List[int]) -> int:
        dp = [[-1]*(2) for _ in range(len(nums))]
        return self.f(nums,0,0,dp)

File: 21Jan2024.java
// Problem Link :- https://leetcode.com/problems/house-robber/

class Solution {
    public int rob(int[] nums) {
        return f(nums, 0, 0, new Integer[nums.length][2]);
    }

    public int f(int[] a, int i, int j, Integer[][] dp) {
        if (i >= a.length)
            return 0;

        if (dp[i][j] != null)
            return dp[i][j];
        int f1 = f(a, i + 1, 0, dp);
        int ans = f1;
        if (j == 0) {
            int f2 = f(a, i + 1, 1, dp) + a[i];
            ans = Math.max(ans, f2);
        }
        return dp[i][j] = ans;
    }
}